sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/MarathonConstraintParser.java

```diff
@@ -0,0 +1,321 @@
+
+/**
+ * Implements support for generating {@link PlacementRule}s from Marathon-style constraint strings.
+ *
+ * @see https://mesosphere.github.io/marathon/docs/constraints.html
+ */
+public class MarathonConstraintParser {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MarathonConstraintParser.class);
+
+    private static final String HOSTNAME_FIELD = "hostname";
+    private static final Map<String, Operator> SUPPORTED_OPERATORS = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+    static {
+        SUPPORTED_OPERATORS.put("UNIQUE", new UniqueOperator());
+        SUPPORTED_OPERATORS.put("CLUSTER", new ClusterOperator());
+        SUPPORTED_OPERATORS.put("GROUP_BY", new GroupByOperator());
+        SUPPORTED_OPERATORS.put("LIKE", new LikeOperator());
+        SUPPORTED_OPERATORS.put("UNLIKE", new UnlikeOperator());
+        SUPPORTED_OPERATORS.put("MAX_PER", new MaxPerOperator());
+    }
+
+    private MarathonConstraintParser() {
+        // do not instantiate
+    }
+
+    /**
+     * ANDs the provided marathon-style constraint string onto the provided hard-coded
+     * {@link PlacementRule}, or returns the provided {@link PlacementRule} as-is if the
+     * marathon-style constraint is {@code null} or empty.
+     *
+     * @throws IOException if {@code marathonConstraints} couldn't be parsed, or if the parsed
+     *     content isn't valid or supported
+     */
+    public static PlacementRule parseWith(PlacementRule rule, String marathonConstraints)
+            throws IOException {
+        PlacementRule marathonRule = parse(marathonConstraints);
+        if (marathonRule instanceof PassthroughRule) {
+            return rule; // pass-through original rule
+        }
+        return new AndRule(rule, marathonRule);
+    }
+
+    /**
+     * Creates and returns a new {@link PlacementRule} against the provided marathon-style
+     * constraint string. Returns a {@link PassthroughRule} if the provided constraint string is
+     * {@code null} or empty.
+     *
+     * @param marathonConstraints the marathon-style constraint string, containing one or more
+     *     nested json list entries of the form {@code [["multi","list","value"],["hello","hi"]]},
+     *     or one or more colon-separated entries of the form {@code multi:list:value,hello:hi},
+     *     or a {@code null} or empty value if no constraint is defined
+     * @throws IOException if {@code marathonConstraints} couldn't be parsed, or if the parsed
+     *     content isn't valid or supported
+     */
+    public static PlacementRule parse(String marathonConstraints) throws IOException {
+        if (marathonConstraints == null || marathonConstraints.isEmpty()) {
+            // nothing to enforce
+            return new PassthroughRule();
+        }
+        List<List<String>> rows = splitConstraints(marathonConstraints);
+        if (rows.size() == 1) {
+            return parseRow(rows.get(0));
+        } else {
```

Comment:
Matter of preference, but you can omit the `else` clause and move that logic into the parent scope given that control flow never gets here if `rows.size() == 1`.